Firestore/core/src/firebase/firestore/immutable/array_sorted_map.h

```diff
@@ -0,0 +1,306 @@
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_ARRAY_SORTED_MAP_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_ARRAY_SORTED_MAP_H_
+
+#include <assert.h>
+
+#include <algorithm>
+#include <array>
+#include <functional>
+#include <memory>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/immutable/map_entry.h"
+
+namespace firebase {
+namespace firestore {
+namespace immutable {
+
+/**
+ * A base class for implementing ArraySortedMap, which contains constants and
+ * types that don't depend upon the type of ArraySortedMap's type parameters.
+ */
+class ArraySortedMapBase {
+ public:
+  /**
+   * The type of size(). Note that this is not size_t specifically to save
+   * space in the TreeSortedMap implementation.
+   */
+  typedef uint32_t size_type;
+
+  /**
+   * The maximum size of an ArraySortedMap.
+   *
+   * This is the size threshold where we use a tree backed sorted map instead
+   * of an array backed sorted map. This is a more or less arbitrary chosen
+   * value, that was chosen to be large enough to fit most of object kind of
+   * Firebase data, but small enough to not notice degradation in performance
+   * for inserting and lookups. Feel free to empirically determine this
+   * constant, but don't expect much gain in real world performance.
+   */
+  static constexpr size_type kFixedSize = 25;
+};
+
+/**
+ * A thin wrapper around std::array that also maintains a dynamic size.
+ *
+ * ArraySortedMap does not actually contain its array: it contains a
+ * shared_ptr to a FixedArray
+ *
+ * @tparam T The contents of an element in the array.
+ * @tparam fixed_size the fixed size to use in creating the FixedArray.
+ */
+template <typename T, ArraySortedMapBase::size_type fixed_size>
+class FixedArray {
+ public:
+  typedef ArraySortedMapBase::size_type size_type;
+  typedef std::array<T, fixed_size> array_type;
+  typedef typename array_type::iterator iterator;
+  typedef typename array_type::const_iterator const_iterator;
+
+  /**
+   * Appends to this array, copying from the given src_begin up to but not
+   * including the src_end.
+   */
+  template <typename SourceIterator>
+  void append(SourceIterator src_begin, SourceIterator src_end) {
+    auto copy_end = std::copy(src_begin, src_end, end());
+    size_ = static_cast<size_type>(copy_end - begin());
+  }
+
+  /**
+   * Appends a single value to the array.
+   */
+  void append(T&& value) {
+    *end() = std::forward<T>(value);
+    size_ += 1;
+  }
+
+  iterator begin() {
+    return contents_.begin();
+  }
+
+  const_iterator begin() const {
+    return contents_.begin();
+  }
+
+  iterator end() {
+    return begin() + size_;
+  }
+
+  const_iterator end() const {
+    return begin() + size_;
+  }
+
+ private:
+  array_type contents_;
+  size_type size_;
+
+  template <typename K, typename V, typename C>
+  friend class ArraySortedMap;
```

Comment:
See some of my comments in ArraySortedMap; perhaps this can be removed.

Reply:
Done.